idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,248 @@
+        tee(bool, optional)            : Indicates that ef solver output should be teed
+        solver_options(dict, optional) : Provides options to the solver (also the name of an attribute)
+
+     Returns:
+         tuple   : results object containing the all information including
+            - results.gradient_f(numpy.ndarray)    : Gradient vector of the objective function with respect to the (decision variables, parameters) at the optimal solution
+            - results.gradient_c(numpy.ndarray)    : Gradient vector of the constraints with respect to the (decision variables, parameters) at the optimal solution. Each row contains [column number, row number, and value], colum order follows variable order in col. If no constraint exists, return []
+            - results.dsdp(numpy.ndarray)          : Gradient vector of the (decision variables, parameters) with respect to paramerters (=theta_name). number of rows = len(theta_name), number of columns= len(col)
+            - results.propagation_c(numpy.ndarray) : Error propagation in the constraints (only constraints have theta_name)  e.g) [[constraint iumber1, error1],[constraint number2, error2]]             
+            - results.propagation_f(numpy.float64) : Error propagation in the objective function, df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+            - results.col(list)                    : List of variable names  
+    Raises:
+        Exception:  When model_uncertain is neither 'ConcreteModel' nor 'function'.
+    """
+
+    # define Pyomo model
+    if type(model_uncertain).__name__ == 'ConcreteModel':
+        model = model_uncertain
+    elif type(model_uncertain).__name__ == 'function':
+        model = model_uncertain()
+    else:
+        raise TypeError('model_uncertain must be either python function or Pyomo ConcreteModel.')
+    # Remove all "'" in theta_names     
+    theta_names, var_dic,variable_clean = clean_variable_name(theta_names)
+    for v in theta_names:
+        eval('model.'+var_dic[v]).setlb(theta[v])
+        eval('model.'+var_dic[v]).setub(theta[v])
+    # get gradient of the objective function, constraints, and the column number of each theta
+    gradient_f, gradient_c, col, line_dic = get_dfds_dcds(model, theta_names, tee)
+    dsdp, col  = get_dsdp(model, theta_names, theta, var_dic,tee)      
+    
+    gradient_f_dic = {}
+    for i in range(len(col)):
+        gradient_f_dic["d(f)/d("+col[i]+")"] = gradient_f[i]
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True)))
+    gradient_c_dic = {}
+    if num_constraints > 0 :
```

Comment:
Do we need `gradient_c_dic` anymore? Why not just use a sparse numpy array? Also, when creating the sparse numpy array, add a comment with the expected dimensions. For example, would be an Ncon by Ntheta matrix where Ncon is the numpy of constraints and Ntheta is the number of uncertainty parameters?

Reply:
I updated them.

Reply:
@JanghoPark-LBL With this update, is there any reason to keep `gradient_c_dic`? Can we remove it?

Reply:
@adowling2 Forgot to remove it. I removed the related lines.